sound/soc/sof/ipc4-topology.c

```diff
@@ -1940,6 +2002,58 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return sof_ipc4_update_hw_params(sdev, pipeline_params, out_audio_fmt);
 }
 
+static int sof_ipc4_prepare_asrc_module(struct snd_sof_widget *swidget,
+					struct snd_pcm_hw_params *fe_params,
+					struct snd_sof_platform_stream_params *platform_params,
+					struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct sof_ipc4_asrc *asrc = swidget->private;
+	struct sof_ipc4_available_audio_format *available_fmt = &asrc->available_fmt;
+	struct sof_ipc4_audio_format *out_audio_fmt;
+	struct sof_ipc4_audio_format *in_audio_fmt;
+	u32 out_ref_rate, out_ref_channels, out_ref_valid_bits;
+	int output_format_index, input_format_index;
+
+	input_format_index = sof_ipc4_init_input_audio_fmt(sdev, swidget, &asrc->base_config,
+							   pipeline_params, available_fmt);
+	if (input_format_index < 0)
+		return input_format_index;
+	/*
+	 * ASRC does not perform format conversion, so the output channels and valid bit depth must
+	 * be the same as that of the input.
+	 */
+	in_audio_fmt = &available_fmt->input_pin_fmts[input_format_index].audio_fmt;
+	out_ref_channels = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(in_audio_fmt->fmt_cfg);
+	out_ref_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(in_audio_fmt->fmt_cfg);
+
+	/*
+	 * For capture, the ASRC module should convert the rate to match the rate requested by the
+	 * PCM hw_params. Set the reference params based on the fe_params unconditionally as it
+	 * will be ignored for playback anyway.
```

Comment:
this is very confusing.

If the input is a nominal 48kHz, but with a different clock master, and the output is also a nominal 48kHz defined by hw_params, then how would ASRC know what to do? The nominal rates are identical on input and output...

Reply:
The 48k to 48k is a quite typical case. The ASRC in asynchronous mode calculates the skew factor for the measured source and sink rates and converts with rate factor that is very close to 1, e.g. 0.999 ... 1.001. Another typical usage is to connect  FW sychronous 48 kHz domain to other follower-DAI 44.1 kHz domain.

It looks the comment text is from SRC prepare but this function has differences. SRC is typically close to host copier while ASRC is close to DAI copier. Does that cause the differences?  Would ASRC not work similar formats logic as SRC?  I don't know, so asking. What do you think @ranj063 ?

Reply:
For this comment to be correct, dont we need the same check as in the SRC prepare to make sure that there's exactly 1 output format for the ASRC module as well?

Reply:
Yes, I think they should be similar.

To add, for 48k-48k ASRC playback both sink and source rates are set by topology and kernel exactly to  48000, asynchronous mode true, operation mode push. For capture direction the same but operation mode pull. ASRC is internally handling the even 48 kHz data flow to other FW pipelines side, and uneven data flow to DAI copier side. If e.g. the real DAI rate is 47990 Hz, then the DAI copier will process to/from ASRC mostly 48 frame periods but sometimes a shorter 47 frame period. ASRC is doing internally the slight waveform "stretch/shrink" and internally all other needed buffering.